script/test-in-container.sh

```diff
@@ -0,0 +1,78 @@
+#
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+testcase=$1
+set -e pipefail
+
+[ -n "$testcase" ] || (echo No testcase provided; exit 1) >&2
+[ -f "$testcase" ] || (echo Cannot find file "$testcase"; exit 1 ) >&2
+
+thisdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+circlecidir="$thisdir"/../.circleci
+[ -d "$circlecidir" ] || circlecidir="$thisdir"/../../../.circleci
+[ -d "$circlecidir" ] || (echo Cannot find .circleci directory; exit 1 ) >&2
+basename=$(basename "$testcase")
+containername="localtest.$basename"
+
+# shellcheck source=/dev/null
+source "$thisdir/osht.sh"
+[[ ${BASH_SOURCE[0]} != *privileged.sh ]] || SKIP test "${PRIVILEGED_TESTS}" != 1 # PRIVILEGED_TESTS is not set to 1
+SKIP test "$(docker info 2>&1)" == "" # Docker doesn't seem to be running
+PLAN 1
+
+# let's use `docker build` here to untilize docker cache
+( 
+# shellcheck disable=SC2046 disable=SC2005
+echo "FROM registry.opensuse.org/devel/openqa/ci/containers/base:latest
+RUN sudo zypper ar -f -p 80 https://download.opensuse.org/repositories/devel:/openQA:/Leap:/15.1/openSUSE_Leap_15.1 openQA
+RUN sudo zypper ar -f -p 81 http://download.opensuse.org/repositories/devel:openQA/openSUSE_Leap_15.1 devel
+RUN sudo zypper --gpg-auto-import-keys ref
+RUN sudo zypper -n install $( echo $( (cat "$circlecidir"/dependencies.txt) | sed -e 's/\r//' |sort))
+RUN sudo zypper -n install apparmor-profiles apparmor-utils"
+) | docker build -t "$containername" -f- "$thisdir"
+
+privileged_option=""
+[[ ${BASH_SOURCE[0]} != *privileged.sh ]] || privileged_option=--privileged
+docker run $privileged_option --rm --name "$containername" -v "$circlecidir/../":/opt/testing_area --entrypoint="/usr/bin/tail" -- "$containername" -f /dev/null &
+
+function cleanup {
+    docker stop -t 0 "$containername" >&/dev/null || :
+    _osht_cleanup
+}
+
+trap cleanup INT TERM EXIT
+counter=0
+
+# wait container start
+until [ $counter -gt 10 ]; do
+  sleep 0.5
+  docker exec "$containername" pwd >& /dev/null && break
+  ((counter++))
+done
+
+docker exec "$containername" pwd >& /dev/null || (echo Cannot start container; exit 1 ) >&2
+
+if [ -n "$CIRCLECI" ]; then # circleci cannot mount volumes, so just copy current dir
+    docker cp "$thisdir"/../. "$containername:/opt/testing_area/"
+fi
+
+set +e
+docker exec -i "$containername" bash < "$testcase"
```

Comment:
Why can't we just execute the testcase with `docker run` directly? These `docket exec` commands and the "sleep loop" seem needlessly complicating the test setup.

Reply:
I believe this way it is easier to distinguish between 'test setup' and 'test run' (thus easier to troubleshoot eventual falures), plus it allows to mix commands in container and on host, the same way as it is done in `.circleci/build_dependencies.sh`

Reply:
I get your point but still don't think it is worth the benefit. But it isn't very important to me so keep it the way you like.

Reply:
I agree with @Martchus here, I would also prefer with a synchronous `docker run`

Reply:
One additional problem with `docker run bash -c "script"` is that it brings a challenge to pass variables and escape special characters properly.
And one more problem is that it is impossible to return into the container once `docker run` is finished, while splitting `run` and `exec` will give a lot of freedom into how container is used.
Last note is that I am working to Docker test which allows using systemd commands and it requires `docker run $image /usr/lib/systemd/systemd` and once the container is up - it is possible to `docker exec $container systemctl start myservice`, which will open path to test systemd scripts and more advanced scenarios locally.

Reply:
sounds great!